pkg/sql/opt/xform/scan_index_iter.go

```diff
@@ -300,6 +334,40 @@ func (it *scanIndexIter) filtersImplyPredicate(
 	return nil, false
 }
 
+// buildConstProjectionsFromPredicate builds a ProjectionsExpr that projects
+// constant values for the given constCols. The constant values are extracted
+// from the given partial index predicate expression. Panics if a constant value
+// cannot be extracted from pred for any of the constCols.
+func (it *scanIndexIter) buildConstProjectionsFromPredicate(
+	pred memo.FiltersExpr, constCols opt.ColSet,
+) memo.ProjectionsExpr {
+	proj := make(memo.ProjectionsExpr, 0, constCols.Len())
+	for col, ok := constCols.Next(0); ok; col, ok = constCols.Next(col + 1) {
+		ord := it.tabMeta.MetaID.ColumnOrdinal(col)
+		typ := it.tabMeta.Table.Column(ord).DatumType()
+
+		val := memo.ExtractValueForConstColumn(pred, it.evalCtx, col)
+		if val == nil {
+			panic(errors.AssertionFailedf("could not extract constant value for column %d", col))
+		}
+
+		var scalar opt.ScalarExpr
+		if val == tree.DNull {
+			// NULL values should always be a memo.NullExpr, not a
+			// memo.ConstExpr.
+			scalar = memo.NullSingleton
```

Comment:
I'm not sure if it's safe to project an untyped `NULL` here, but I did some simply ad-hoc testing and it seemed fine.